### DISCUSSION TITLE
should we explore letting users mix raw Envoy config into the Contour-generated output?

### DISCUSSION THREAD

**Discussion by skriss** [projectcontour/contour#3188] [Ideas]
One way to give users more flexibility to customize the Envoy config without having to hand-code every desired tuneable would be to give them some way to mix raw Envoy config into Contour's generated output. 

There are obviously plenty of complexities and pitfalls here so I'm not saying we definitely should do it; I'm just wondering if folks think it's something we should explore more, or whether it's a non-starter.

**Comment by jpeach** [comment]
Some related issues are https://github.com/projectcontour/contour/issues/2385, https://github.com/projectcontour/contour/issues/3006. Istio also has a [CRD](https://istio.io/latest/docs/reference/config/networking/envoy-filter/) for patching Envoy config.

**Comment by skriss** [comment]
Also found https://github.com/projectcontour/contour/issues/857 which has some interesting back-story and discussion.

**Comment by youngnick** [comment]
The problem here is that if we allow people do to this, how do we support them when it breaks? How do we ensure that people mixing in Envoy things won't inadvertently create a security problem (such as with the SNI routing problem that we had to do an inline Lua fix for?)

I agree that just giving people this would one hundred percent make a large class of issues go away, and people have asked for it specifically in the past. But is the tradeoff worth it? I'm not convinced it is.

What's Istio's stance on support for if you use the patching?

**Reply by jpeach** [reply]
> The problem here is that if we allow people do to this, how do we support them when it breaks?

Yeh I think that we would want to provide substantially better debugging tools for this.
